### ISSUE TITLE
Suggestion: add an 'event' argument to onAny handlers

### ISSUE THREAD

**Issue by fresheneesz** [asyncly/EventEmitter2#139]
There's a problem with using `this.event` to get the event. If you're in the habit of binding a this onto handlers, you'll be unable to do this for onAny handlers. For example:

```
emitter.onAny(function() {
    // this.event is almost definitely undefined, and definitely isn't the triggering event 
}.bind(this))
```

I suggest that instead, the onAny API be changed to this:

```
emitter.onAny(function(event, argument1/*, ...*/) {
   console.log("Got a "+event+" event!")
}.bind(this))

emitter.emit("whatever", 4)
```

**Comment by hij1nx** [comment]
Open to a PR that makes the first arg the event.

On Friday, February 20, 2015, fresheneesz <notifications@github.com> wrote:

> There's a problem with using this.event to get the event. If you're in
> the habit of binding a this onto handlers, you'll be unable to do this for
> onAny handlers. For example:
>
> emitter.onAny(function() {
>     // this.event is almost definitely undefined, and definitely isn't the triggering event
> }.bind(this))
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/asyncly/EventEmitter2/issues/139>.
>


-- 
Sent from Gmail Mobile

**Comment by fresheneesz** [comment]
@hij1nx Great, that's much clearer! The only question I have is what happens when those methods are called on a range that already has some selection associated with it? Does it completely change the selection, or does it just "grow" the selection?

**Comment by cxreg** [comment]
I don't see any obvious reason why this.event can't be set before invoking _all handlers.  What am I missing?

**Comment by fresheneesz** [comment]
Well antipattern or not, I've seen lots of code use this pattern, and have used it myself (tho I usually create a `that` variable i assign to `this` i can then use in the callback). The issue is that this is the only thing that causes `EventEmitter2` to not be a fully compatible drop-in replacement for node's EventEmitter, which is a good feature to have for a library that already so closely follows the same API. And `emitter.event` should never contain the event - this would break things in cases where you emit another event from the emitter inside the event handler, and then lower in the event handler access emitter.event.

**Comment by RangerMauve** [comment]
@cxreg, I'm not seeing the re-binding in the PR code. What file/line is it in?

**Comment by fresheneesz** [comment]
@RangerMauve I think he meant the rebinding in my example in the original post up at the top. The `bind(this)`.

**Comment by Yvem** [comment]
+1

**Comment by Yvem** [comment]
It seems this is implemented. I see this.event correctly set in my EventEmitter2.onAny handler. Issue to be closed ?

**Comment by RangerMauve** [comment]
@Yvem `this.event` is being set, but the PR adds the `event` as the first argument of the callback rather than a property of `this`. The reason being is that functions that got bound to a different `this` context cannot access the event name in the current implementation.

**Comment by ghostsquad** [comment]
+1

**Comment by RangerMauve** [comment]
Fixed in #141

**Comment by jessedijkstra** [comment]
:+1:
